presto-hive/src/main/java/com/facebook/presto/hive/util/AsyncQueue.java

```diff
@@ -13,103 +13,111 @@
-    private final AtomicBoolean finished = new AtomicBoolean();
-    private final AtomicReference<CompletableFuture<?>> notEmptyFuture = new AtomicReference<>(new CompletableFuture<>());
-
-    /**
-     * Adds an element to the queue.  Elements added after the queue is finished
-     * are normally ignored (without error), but if someone calls finish while
-     * {@code add} is running the element may still be added.
-     *
-     * @throws NullPointerException if the element is null
-     */
-    public void add(T element)
+    public static final CompletableFuture<?> COMPLETED_FUTURE = CompletableFuture.completedFuture(null);
+
+    private final int softMaxQueueSize;
+
+    private final BlockingQueue<T> elements = new LinkedBlockingQueue<>();
+    // Upper bound of number of elements in `elements`. Increment happens before add. Decrement happens after removal.
+    // The order is chosen to guarantee that this variable never becomes negative.
+    private final AtomicInteger queueSize = new AtomicInteger();
+    private final AtomicReference<CompletableFuture<?>> offerResubmissionFutureReference = new AtomicReference<>(new CompletableFuture<>());
+    private final AtomicReference<CompletableFuture<?>> fetchResubmissionFutureReference = new AtomicReference<>(new CompletableFuture<>());
+
+    public AsyncQueue(int softMaxQueueSize)
+    {
+        this.softMaxQueueSize = softMaxQueueSize;
+    }
+
+    public int size()
     {
-        requireNonNull(element, "element is null");
-        if (finished.get()) {
-            return;
+        return queueSize.get();
+    }
+
+    public CompletableFuture<?> offer(T element)
+    {
+        // Make sure to grab the futures before adding to the queue.
+        // This guarantees that the future will be current (or already signaled), but never newer.
+        CompletableFuture<?> offerResubmissionFuture = offerResubmissionFutureReference.get();
+
+        int newSize = queueSize.incrementAndGet();
+        elements.add(element);
+        if (newSize == 1) {
+            signalFetchResubmissionAndReset();
         }
-        queue.add(element);
-        notEmptyFuture.get().complete(null);
+        return newSize < softMaxQueueSize ? COMPLETED_FUTURE : offerResubmissionFuture;
     }
 
-    /**
-     * Gets a batch of elements from the queue.  If queue is not empty, a
-     * completed future containing the queued elements is returned.  If the
-     * queue is empty, an incomplete future is returned that completes when
-     * an element is added.
-     * <p>
-     * It is possible that an empty list will be returned as another reader
-     * may have consumed all of the data.  A caller should read until, the
-     * {@code isFinished} method returns {@code true}.
-     *
-     * @param maxSize the maximum number of elements to return.
-     */
-    public CompletableFuture<List<T>> getBatchAsync(int maxSize)
+    private int drainTo(Collection<T> collection, int maxSize)
```

Comment:
I agree that the code is easier to read if all these methods that are only used once are inlined.